nixos/modules/services/web-servers/tomcat/default.nix

```diff
@@ -0,0 +1,480 @@
+        workDir = "${cacheDir}/${hostOpts.hostName}";
+      };
+    in
+      ''
+        <Host ${concatStringsSep " " (mapAttrsToList (k: v: "${k}=\"${toStr v}\"") (defaultAttrs // hostOpts.settings))}>
+          ${concatMapStrings (alias: "<Alias>${alias}</Alias>") hostOpts.serverAliases}
+          ${acmeChallenge}
+          ${robotsEntries}
+          <Valve
+            className="org.apache.catalina.valves.AccessLogValve"
+            directory="${logsDir}"
+            prefix="${hostOpts.hostName}_access" suffix=".log"
+            pattern="%h %l %u %t &quot;%r&quot; %s %b"
+          />
+          ${hostOpts.extraConfig}
+        </Host>
+      ''
+  ;
+
+  serverXmlFile = pkgs.writeText "server.xml" ''
+    <?xml version="1.0" encoding="UTF-8"?>
+    <!--
+      This is a generated file.  Do not edit!
+
+      To make changes, edit the services.tomcat NixOS options
+      in your /etc/nixos/configuration.nix file.
+    -->
+    <Server ${concatStringsSep " " (mapAttrsToList (k: v: "${k}=\"${toStr v}\"") cfg.serverAttrs)}>
+      ${cfg.extraConfig}
+
+      <Service name="Catalina">
+        ${concatMapStrings mkConnectorConf cfg.connectors}
+
+        <Engine ${concatStringsSep " " (mapAttrsToList (k: v: "${k}=\"${toStr v}\"") cfg.engineAttrs)}>
+          ${concatMapStrings mkVHostConf vhosts}
+        </Engine>
+      </Service>
+    </Server>
+  '';
+
+  # TODO: write a custom logging.properties file i guess
+  # https://github.com/GoogleCloudPlatform/tomcat-runtime/pull/28/files
+  loggingProperties = pkgs.writeText "logging.properties" ''
+    handlers = java.util.logging.ConsoleHandler
+
+    .handlers = java.util.logging.ConsoleHandler
+
+    ############################################################
+    # Handler specific properties.
+    # Describes specific configuration info for Handlers.
+    ############################################################
+
+    java.util.logging.ConsoleHandler.level = FINE
+    java.util.logging.ConsoleHandler.formatter = org.apache.juli.OneLineFormatter
+
+    ############################################################
+    # Facility specific properties.
+    # Provides extra control for each logger.
+    ############################################################
+
+    org.apache.catalina.core.ContainerBase.[Catalina].[localhost].level = INFO
+    org.apache.catalina.core.ContainerBase.[Catalina].[localhost].handlers = java.util.logging.ConsoleHandler
+
+    org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/manager].level = INFO
+    org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/manager].handlers = java.util.logging.ConsoleHandler
+
+    org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/host-manager].level = INFO
+    org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/host-manager].handlers = java.util.logging.ConsoleHandler
+  '';
+in
+{
+  meta = {
+    maintainers = with maintainers; [ danbst ];
```

Comment:
this can be removed, I no longer maintain Tomcat installations.

Reply:
OK.